File: prd-api/src/PrdAgent.Core/Models/Skill.cs
namespace PrdAgent.Core.Models;

/// <summary>
/// 统一技能模型（替代 PromptEntry + SkillEntry）
/// - 每个文档是一个独立技能，存储在 skills 集合
/// - 系统技能由管理员创建（visibility: system），不可删除
/// - 公共技能由管理员创建（visibility: public），可被所有用户看到
/// - 个人技能由用户创建（visibility: personal），仅创建者可见
/// </summary>
public class Skill
{
    public string Id { get; set; } = string.Empty;

    /// <summary>唯一标识（kebab-case）</summary>
    public string SkillKey { get; set; } = string.Empty;

    /// <summary>技能名称</summary>
    public string Title { get; set; } = string.Empty;

    /// <summary>技能描述</summary>
    public string Description { get; set; } = string.Empty;

    /// <summary>图标（emoji）</summary>
    public string? Icon { get; set; }

    /// <summary>分类</summary>
    public string Category { get; set; } = "general";

    /// <summary>标签</summary>
    public List<string> Tags { get; set; } = new();

    // === 可见性 ===

    /// <summary>可见性: system | public | personal</summary>
    public string Visibility { get; set; } = SkillVisibility.Public;

    /// <summary>个人技能的创建者（system/public 为 null）</summary>
    public string? OwnerUserId { get; set; }

    /// <summary>适用角色（空 = 全部角色可用）</summary>
    public List<UserRole> Roles { get; set; } = new();

    /// <summary>排序号</summary>
    public int Order { get; set; }

    // === 输入配置 ===

    /// <summary>输入配置</summary>
    public SkillInputConfig Input { get; set; } = new();

    // === 执行配置（仅服务端持有，不下发客户端） ===

    /// <summary>执行配置</summary>
    public SkillExecutionConfig Execution { get; set; } = new();

    // === 输出配置 ===

    /// <summary>输出配置</summary>
    public SkillOutputConfig Output { get; set; } = new();

    // === 元数据 ===

    /// <summary>是否启用</summary>
    public bool IsEnabled { get; set; } = true;

    /// <summary>是否为系统内置（不可删除）</summary>
    public bool IsBuiltIn { get; set; }

    /// <summary>使用次数</summary>
    public int UsageCount { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>技能可见性常量</summary>
public static class SkillVisibility
{
    /// <summary>系统内置（管理员管理，所有人可见）</summary>
    public const string System = "system";
    /// <summary>公共（管理员创建，所有人可见）</summary>
    public const string Public = "public";
    /// <summary>个人（用户自建，仅自己可见）</summary>
    public const string Personal = "personal";
}

/// <summary>技能输入配置</summary>
public class SkillInputConfig
{
    /// <summary>上下文范围: all | current | prd | none</summary>
    public string ContextScope { get; set; } = "prd";

    /// <summary>是否接受用户附加文本输入</summary>
    public bool AcceptsUserInput { get; set; }

    /// <summary>用户输入占位提示</summary>
    public string? UserInputPlaceholder { get; set; }

    /// <summary>是否接受附件</summary>
    public bool AcceptsAttachments { get; set; }

    /// <summary>可配置参数</summary>
    public List<SkillParameter> Parameters { get; set; } = new();
}

/// <summary>技能执行配置（服务端专有，不下发客户端）</summary>
public class SkillExecutionConfig
{
    /// <summary>提示词模板（支持 {{变量}} 占位符）</summary>
    public string PromptTemplate { get; set; } = string.Empty;

    /// <summary>系统提示词覆盖（null = 使用默认角色系统提示词）</summary>
    public string? SystemPromptOverride { get; set; }

    /// <summary>LLM Gateway 路由标识</summary>
    public string? AppCallerCode { get; set; }

    /// <summary>模型类型偏好</summary>
    public string ModelType { get; set; } = "chat";

    /// <summary>期望模型提示</summary>
    public string? ExpectedModel { get; set; }

    /// <summary>后处理工具链</summary>
    public List<SkillToolStep> ToolChain { get; set; } = new();
}

/// <summary>工具链步骤</summary>
public class SkillToolStep
{
    /// <summary>工具标识: chat, download, clipboard, create-defect 等</summary>
    public string ToolKey { get; set; } = string.Empty;

    /// <summary>工具专有配置</summary>
    public Dictionary<string, object> Config { get; set; } = new();

    /// <summary>是否可选（失败不中断流程）</summary>
    public bool Optional { get; set; }
}

/// <summary>技能输出配置</summary>
public class SkillOutputConfig
{
    /// <summary>输出模式: chat | download | clipboard</summary>
    public string Mode { get; set; } = "chat";

    /// <summary>下载模式的文件名模板</summary>
    public string? FileNameTemplate { get; set; }

    /// <summary>下载模式的 MIME 类型</summary>
    public string? MimeType { get; set; }

    /// <summary>非 chat 模式是否同时在对话中回显</summary>
    public bool EchoToChat { get; set; }
}

/// <summary>技能执行请求</summary>
public class SkillExecuteRequest
{
    /// <summary>会话 ID</summary>
    public string SessionId { get; set; } = string.Empty;

    /// <summary>用户附加文本</summary>
    public string? UserInput { get; set; }

    /// <summary>附件 ID 列表</summary>
    public List<string>? AttachmentIds { get; set; }

    /// <summary>参数覆盖</summary>
    public Dictionary<string, string>? Parameters { get; set; }

    /// <summary>上下文范围覆盖（null = 使用技能默认值）</summary>
    public string? ContextScopeOverride { get; set; }

    /// <summary>输出模式覆盖（null = 使用技能默认值）</summary>
    public string? OutputModeOverride { get; set; }
}

File: prd-desktop/src-tauri/src/commands/skill.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::command;

use crate::models::ApiResponse;
use crate::services::ApiClient;

// Re-export for dialog access
use tauri_plugin_dialog::DialogExt;

// ━━━ 新 Skill API 模型（对应 /api/prd-agent/skills） ━━━━━━━━

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillInputConfig {
    pub context_scope: String,
    pub accepts_user_input: bool,
    pub user_input_placeholder: Option<String>,
    pub accepts_attachments: bool,
    pub parameters: Vec<SkillParameter>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillOutputConfig {
    pub mode: String,
    pub file_name_template: Option<String>,
    pub echo_to_chat: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillParameter {
    pub key: String,
    pub label: String,
    #[serde(rename = "type")]
    pub param_type: String,
    pub default_value: Option<String>,
    pub options: Option<Vec<SkillParameterOption>>,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillParameterOption {
    pub value: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Skill {
    pub skill_key: String,
    pub title: String,
    pub description: String,
    pub icon: Option<String>,
    pub category: String,
    pub tags: Vec<String>,
    pub roles: Vec<String>,
    pub order: i32,
    pub visibility: String,
    pub input: SkillInputConfig,
    pub output: SkillOutputConfig,
    pub is_enabled: bool,
    pub is_built_in: bool,
    pub usage_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillsResponse {
    pub skills: Vec<Skill>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillExecuteResponse {
    pub run_id: String,
    pub user_message_id: String,
    pub assistant_message_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillExecuteRequest {
    pub session_id: String,
    pub user_input: Option<String>,
    pub attachment_ids: Option<Vec<String>>,
    pub parameters: Option<HashMap<String, String>>,
    pub context_scope_override: Option<String>,
    pub output_mode_override: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSkillRequest {
    pub title: String,
    pub description: Option<String>,
    pub icon: Option<String>,
    pub category: Option<String>,
    pub tags: Option<Vec<String>>,
    pub order: Option<i32>,
    pub input: Option<SkillInputConfig>,
    pub execution: Option<SkillExecutionConfig>,
    pub output: Option<SkillOutputConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillExecutionConfig {
    pub prompt_template: String,
    pub system_prompt_override: Option<String>,
    pub model_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSkillResponse {
    pub skill_key: String,
}

/// 获取可用技能列表（新 API：/api/prd-agent/skills）
#[command]
pub async fn get_skills(role: Option<String>) -> Result<ApiResponse<SkillsResponse>, String> {
    let client = ApiClient::new();
    let mut path = "/api/prd-agent/skills".to_string();
    if let Some(r) = role {
        let r = r.trim().to_string();
        if !r.is_empty() {
            path = format!("/api/prd-agent/skills?role={}", r);
        }
    }
    client.get(&path).await
}

/// 执行技能（创建 SkillRun）
#[command]
pub async fn execute_skill(
    skill_key: String,
    session_id: String,
    user_input: Option<String>,
    attachment_ids: Option<Vec<String>>,
    parameters: Option<HashMap<String, String>>,
) -> Result<ApiResponse<SkillExecuteResponse>, String> {
    let client = ApiClient::new();
    let request = SkillExecuteRequest {
        session_id,
        user_input,
        attachment_ids,
        parameters,
        context_scope_override: None,
        output_mode_override: None,
    };
    client
        .post(
            &format!("/api/prd-agent/skills/{}/execute", skill_key),
            &request,
        )
        .await
}

/// 创建个人技能
#[command]
pub async fn create_skill(
    request: CreateSkillRequest,
) -> Result<ApiResponse<CreateSkillResponse>, String> {
    let client = ApiClient::new();
    client.post("/api/prd-agent/skills", &request).await
}

/// 更新个人技能
#[command]
pub async fn update_skill(
    skill_key: String,
    request: CreateSkillRequest,
) -> Result<ApiResponse<serde_json::Value>, String> {
    let client = ApiClient::new();
    client
        .put(&format!("/api/prd-agent/skills/{}", skill_key), &request)
        .await
}

/// 删除个人技能
#[command]
pub async fn delete_skill(skill_key: String) -> Result<ApiResponse<serde_json::Value>, String> {
    let client = ApiClient::new();
    client
        .delete(&format!("/api/prd-agent/skills/{}", skill_key))
        .await
}

// ━━━ 从消息提炼提示词模板 ━━━━━━━━

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateSkillFromMessageRequest {
    pub user_message: Option<String>,
    pub assistant_message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtractPromptTemplateResponse {
    pub prompt_template: String,
}

/// 从对话消息提炼可复用的提示词模板（旧版：单条消息）
#[command]
pub async fn generate_skill_from_message(
    user_message: Option<String>,
    assistant_message: String,
) -> Result<ApiResponse<ExtractPromptTemplateResponse>, String> {
    let client = ApiClient::new();
    let request = GenerateSkillFromMessageRequest {
        user_message,
        assistant_message,
    };
    client
        .post("/api/prd-agent/skills/generate-from-message", &request)
        .await
}

// ━━━ 从多轮对话提炼技能（增强版） ━━━━━━━━

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConversationMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateSkillFromConversationRequest {
    pub conversation_messages: Vec<ConversationMessage>,
    pub key_assistant_message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtractedSkillDraftResponse {
    pub prompt_template: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub icon: Option<String>,
    pub skill_md: Option<String>,
}

// ━━━ 导出技能为 SKILL.md 文件 ━━━━━━━━

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportSkillResponse {
    pub skill_md: String,
    pub file_name: String,
}

/// 导出技能为 SKILL.md（从 API 获取）
#[command]
pub async fn export_skill(skill_key: String) -> Result<ApiResponse<ExportSkillResponse>, String> {
    let client = ApiClient::new();
    client
        .get(&format!("/api/prd-agent/skills/{}/export", skill_key))
        .await
}

/// 导入 SKILL.md 文本创建个人技能
#[command]
pub async fn import_skill(skill_md: String) -> Result<ApiResponse<CreateSkillResponse>, String> {
    let client = ApiClient::new();
    let body = serde_json::json!({ "skillMd": skill_md });
    client.post("/api/prd-agent/skills/import", &body).await
}

/// 将 SKILL.md 内容保存为本地文件（使用系统保存对话框）
#[command]
pub async fn save_skill_to_file(
    app: tauri::AppHandle,
    content: String,
    default_name: String,
) -> Result<bool, String> {
    use std::sync::mpsc;
    use tauri_plugin_dialog::FilePath;

    let (tx, rx) = mpsc::channel();

    app.dialog()
        .file()
        .set_file_name(&default_name)
        .add_filter("SKILL.md", &["md"])
        .add_filter("All Files", &["*"])
        .save_file(move |path| {
            tx.send(path).ok();
        });

    let path = rx.recv().map_err(|e| format!("Dialog error: {}", e))?;

    match path {
        Some(file_path) => {
            let path_buf = match file_path {
                FilePath::Path(p) => p,
                FilePath::Url(u) => {
                    // Convert file:// URL to path
                    u.to_file_path()
                        .map_err(|_| "Invalid file URL".to_string())?
                }
            };
            std::fs::write(&path_buf, content.as_bytes())
                .map_err(|e| format!("Failed to write file: {}", e))?;
            Ok(true)
        }
        None => Ok(false), // User cancelled
    }
}

/// 从多轮对话提炼可复用的技能草案（增强版）
#[command]
pub async fn generate_skill_from_conversation(
    conversation_messages: Vec<ConversationMessage>,
    key_assistant_message: String,
) -> Result<ApiResponse<ExtractedSkillDraftResponse>, String> {
    let client = ApiClient::new();
    let request = GenerateSkillFromConversationRequest {
        conversation_messages,
        key_assistant_message,
    };
    client
        .post("/api/prd-agent/skills/generate-from-conversation", &request)
        .await
}
